src/Microsoft.OData.Core/Json/BitStack.cs

```diff
@@ -0,0 +1,176 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.OData.Json
+{
+    /// <summary>
+    /// An efficient stack optimized for storing data that has two possible kinds of value, represented
+    /// as true and false.
+    /// </summary>
+    /// <remarks>
+    /// This has been adapted from the .NET runtime's internal BitStack which is used by Utf8JsonWriter
+    /// https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/src/System/Text/Json/BitStack.cs
+    /// This has been slightly modified because the original Pop() method did not return the last value pushed.

```

Comment:
Just curious. What does the original Pop() method return?

Reply:
It seems to return the value of the item before the most recent item, or something like that.